### PR TITLE
New module: Add module to create Zabbix maps based on data written in DOT language. (monitoring/zabbix_map)

### PR DESCRIPTION
**Pull request by Akint**
##### SUMMARY

New module allows to create Zabbix maps based on data written in DOT language. This data could be easily rendered from Jinja2 template and inventory data, for example.
This module uses the same common parameters as existing `zabbix_*` modules. Also shares a dependency to `zabbix-api` Python module.

##### ISSUE TYPE

 - New Module Pull Request

##### COMPONENT NAME

zabbix_map

##### ANSIBLE VERSION

```
ansible 2.8.0.dev0 (zabbix_map 22c28dd997) last updated 2018/11/08 16:32:09 (GMT +000)
  config file = None
  configured module search path = [u'/home/akint/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/akint/git/ansible/lib/ansible
  executable location = /home/akint/git/ansible/bin/ansible
  python version = 2.7.12 (default, Dec  4 2017, 14:50:18) [GCC 5.4.0 20160609]
```

##### ADDITIONAL INFORMATION

Example inventory:
```
[web]
web[01:03].example.com

[db]
db.example.com

[backup]
backup.example.com
```

Contents of `map.j2`:
```
digraph G { 
    graph [layout=dot splines=false overlap=scale]
    INTERNET [zbx_url="Google:https://google.com" zbx_image="Cloud_(96)"]
{% for web_host in groups.web %}
    {% set web_loop = loop %}
    web{{ '%03d' % web_loop.index }} [zbx_host="{{ web_host }}"]
    INTERNET -> web{{ '%03d' % web_loop.index }} [zbx_trigger="{{ web_host }}:Zabbix agent on {HOST.NAME} is unreachable for 5 minutes"]
    {% for db_host in groups.db %}
      {% set db_loop = loop %}
    web{{ '%03d' % web_loop.index }} -> db{{ '%03d' % db_loop.index }}
    {% endfor %}
{% endfor %}
    { rank=same
{% for db_host in groups.db %}
    {% set db_loop = loop %}
    db{{ '%03d' % db_loop.index }} [zbx_host="{{ db_host }}"]
    {% for backup_host in groups.backup %}
        {% set backup_loop = loop %}
        db{{ '%03d' % db_loop.index }} -> backup{{ '%03d' % backup_loop.index }} [color="blue"]
    {% endfor %}
{% endfor %}
{% for backup_host in groups.backup %}
    {% set backup_loop = loop %}
        backup{{ '%03d' % backup_loop.index }} [zbx_host="{{ backup_host }}"]
{% endfor %}
    }   
}
```
Example playbook:
```
- name: Example prerequisites
  hosts: all
  gather_facts: no

  tasks:
  - name: Create Zabbix host
    zabbix_host:
      server_url: http://zabbix.example.com
      login_user: username
      login_password: password
      host_name: "{{ inventory_hostname }}"
      state: present
      host_groups:
        - Linux servers
      interfaces:
        - type: 1
          main: 1
          useip: 1
          ip: 127.0.0.1
          dns: ""
          port: 10050
      link_templates:
        - Template App Zabbix Agent
    delegate_to: localhost

- name: Working with 'zabbix_map' module
  hosts: all
  gather_facts: no

  tasks:
  - name: Create Zabbix map
    zabbix_map:
      server_url: http://zabbix.example.com
      login_user: username
      login_password: password
      name: Demo map
      state: present
      data: "{{ lookup('template', 'map.j2') }}"
      default_image: Server_(64)
      expand_problem: no
      highlight: no
      label_type: label
    delegate_to: localhost
    run_once: yes

  - name: Delete Zabbix map
    zabbix_map:
      server_url: http://zabbix.example.com
      login_user: username
      login_password: password
      name: Demo map
      state: absent
    delegate_to: localhost
    run_once: yes
```

Result:
map

```
Using /etc/ansible/ansible.cfg as config file

PLAY [Example prerequisites] ***************************************************

TASK [zabbix_host] *************************************************************
ok: [web02.example.com -> localhost] => {"changed": false}
ok: [web05.example.com -> localhost] => {"changed": false}
ok: [web03.example.com -> localhost] => {"changed": false}
ok: [web01.example.com -> localhost] => {"changed": false}
ok: [web04.example.com -> localhost] => {"changed": false}
ok: [db.example.com -> localhost] => {"changed": false}
ok: [backup.example.com -> localhost] => {"changed": false}

PLAY [Working with 'zabbix_map' module] ****************************************

TASK [Create Zabbix map] *******************************************************
changed: [web01.example.com -> localhost] => {"changed": true, "result": "Successfully created map: Demo map"}

TASK [Delete Zabbix map] *******************************************************
changed: [web01.example.com -> localhost] => {"changed": true, "result": "Successfully deleted map: Demo map"}

PLAY RECAP *********************************************************************
backup.example.com         : ok=1    changed=0    unreachable=0    failed=0   
db.example.com             : ok=1    changed=0    unreachable=0    failed=0   
web01.example.com          : ok=3    changed=2    unreachable=0    failed=0   
web02.example.com          : ok=1    changed=0    unreachable=0    failed=0   
web03.example.com          : ok=1    changed=0    unreachable=0    failed=0   
web04.example.com          : ok=1    changed=0    unreachable=0    failed=0   
web05.example.com          : ok=1    changed=0    unreachable=0    failed=0   
```

### CODE REVIEW

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 1]
```diff
@@ -0,0 +1,800 @@
...
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]:
```
from __future__ import absolute_import, division, print_function
__metaclass__ = type
```
is missing somewhere around here

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 2]
```diff
@@ -0,0 +1,800 @@
...
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: This copyright notice can probably be condensed to something like in zabbix_host.py

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 3]
```diff
@@ -0,0 +1,800 @@
...
+
+DOCUMENTATION = '''
+---
+module: zabbix_map
+author:
+    - "Anton Alekseyev (@Akint)"
+short_description: Zabbix map creates/updates/deletes
+description:
+    - "This module allows you to create, modify and delete Zabbix map entries, using DOT language.
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: Maybe mention GraphViz here

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 4]
```diff
@@ -0,0 +1,800 @@
...
+        C(zbx_trigger) contains name of the trigger used as a link indicator in C(host_name:trigger_name) format.
+            More than one trigger could be specified by adding some prefix (e.g., C(zbx_trigger1), C(zbx_trigger2)).
+        C(zbx_trigger_color) contains indicator color specified either as CSS3 name or as a hexadecimal code starting with C(#).
+        C(zbx_trigger_draw_style) contains indicator draw style. Possible values are the same as for C(zbx_draw_style)."
+requirements:
+    - "python >= 2.6"
+    - zabbix-api
+    - pydotplus
+    - webcolors
+    - PIL
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: PIL is not available in Python 3, and ansible is compatible with Python 3.5+.

Pillow should be a drop-in replacement.

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 5]
```diff
@@ -0,0 +1,800 @@
...
+    - Graphviz
+version_added: "2.4"
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: now 2.5

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 6]
```diff
@@ -0,0 +1,800 @@
...
+    default_image: Server_(64)
+    expand_problem: no
+    highlight: no
+    label_type: label
+  delegate_to: localhost
+  run_once: yes
+'''
+
+ANSIBLE_METADATA = {
+    'metadata_version': '1.0',
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: Now 1.1

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 7]
```diff
@@ -0,0 +1,800 @@
...
+            return False
+        self._update_ids(generated_map_config, exist_map_config)
+        if not self._is_links_equal(generated_map_config['links'], exist_map_config['links']):
+            return False
+        return True
+
+    def _is_selements_equal(self, generated_selements, exist_selements):
+        if len(generated_selements) != len(exist_selements):
+            return False
+        generated_selements_sorted = sorted(generated_selements, key=itemgetter('elementid', 'label'))
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: elementid no longer exists in the Zabbix 3.4 API on the map object.
Would selementid suffice here?

(this crashes when used with Zabbix 3.4)

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 8]
```diff
@@ -0,0 +1,800 @@
...
+    )
+
+    if not HAS_ZABBIX_API:
+        module.fail_json(msg="Missing required zabbix-api module (check docs or install with: pip install zabbix-api)")
+    if not HAS_PYDOTPLUS:
+        module.fail_json(msg="Missing required pydotplus module (check docs or install with: pip install pydotplus)")
+    if not HAS_WEBCOLORS:
+        module.fail_json(msg="Missing required webcolors module (check docs or install with: pip install webcolors)")
+    if not HAS_PIL:
+        module.fail_json(msg="Missing required PIL module (check docs or install with: pip install PIL)")
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: or pip install Pillow (on Python 3)

File: lib/ansible/modules/monitoring/zabbix_map.py [hunk 9]
```diff
@@ -0,0 +1,800 @@
...
+        default: true
+    label_type:
+        description:
+            - Map element label type.
+        required: false
+        choices: ['label', 'ip', 'name', 'status', 'nothing', 'custom']
+        default: "name"
+'''
+
+RETURN = ''' # '''
```
Comment (eikef) [inline lib/ansible/modules/monitoring/zabbix_map.py]: While return values should be documented for new modules, the other zabbix modules do not document this either (as yet). However, there should not be a need to explicitly define an empty variable.

**Review by eikef — CHANGES_REQUESTED** [review]
Generally this looks good. 

A few updates are required for current ansible requirements (version_added, metadata, imports, metaclass).

The module also does not work with current Zabbix release 3.4; in between 3.2 and 3.4, elementid got removed from the Zabbix API.

Whether Pillow is adequate as a replacement for PIL needs to be tested.

**Comment by D3DeFi** [comment]
Hello @Akint, this PR has been silent for a while now, do you by any chance plan to implement changes suggested by @eikef ?

**Comment by Akint** [comment]
Hi @D3DeFi,

Thank you for pinging me. I have completely missed mentioned suggestions. Will implement them in the next few days.

File: lib/ansible/modules/monitoring/zabbix_map.py (hunk 9 above)
Comment (Akint) [reply lib/ansible/modules/monitoring/zabbix_map.py]: [Documentation](https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_documenting.html#return-block) says: 
> If your module doesn’t return anything (apart from the standard returns), this section of your module should read: ```RETURN = ''' # '''```

**Comment by Akint** [comment]
@eikef, @D3DeFi, I have implemented most of the suggested changes. Could you please take a look?

**Comment by D3DeFi** [comment]
@Akint soo sorry for the delay. I will try to get to your PR in the upcoming days. Thanks once again for implementing requested changes and for being so patient 👍

File: lib/ansible/modules/monitoring/zabbix/zabbix_map.py [hunk 10]
```diff
@@ -0,0 +1,812 @@
...
+            'trigger': 2,
+            'group': 3,
+            'image': 4
+        }
+        element_type = {
+            'elementtype': types['image'],
+            # Elementid was mandatory even for image elements until Zabbix 3.4.
+            # Since Zabbix API ignores unknown fields, this is left here for simplicity:
+            # it works for Zabbix <= 3.2 and doesn't break Zabbix 3.4.
+            'elementid': "0",
```
Comment (D3DeFi) [inline lib/ansible/modules/monitoring/zabbix/zabbix_map.py]: Commenting this seems to fix the error for zabbix-server 3.4

**Review by D3DeFi — CHANGES_REQUESTED** [review]
Very nice module @Akint :) Thank you for the contribution. I've tested this against zbx-server 3.0, 3.2 and 3.4. First two work as intended. Unfortunately, module is still failing when run against zbx-server 3.4 with the following error:
```yaml
'Failed to update map: (u''Error -32500: Application error., No permissions to referred object or it does not exist! while sending
{
    "params": {
        "width": 800,
        "label_type": 0,
        "name": "ansible test",
        "links": [],
        "selements": [
            {
                "elements": [
                    {
                        "hostid": "2"
                    }
                ],
                "y": 268,
                "label": "web001.example.com",
                "iconid_off": "150",
                "selementid": 1,
                "x": 376,
                "elementtype": 0,
                "elementid": "0"    # this here seems to cause the problem
            }
        ],
        "expandproblem": 0,
        "highlight": 0,
        "sysmapid": "3",
        "height": 600
    },
    "jsonrpc": "2.0",
    "method": "map.update",
    "auth": "<<token>>",
    "id": 7
}'', -32500)
```

As for the rest of the module, it looks good to me and I will vote `shipit` when the above mentioned bug is fixed. And please update PR description to include ansible 2.8 version instead of 2.2.

File: lib/ansible/modules/monitoring/zabbix/zabbix_map.py (hunk 10 above)
Comment (Akint) [reply lib/ansible/modules/monitoring/zabbix/zabbix_map.py]: @D3DeFi, thank you for your response! I have fixed the bug, updated PR description and also checked the module against Zabbix 4.0, which seems to be working as well.

**Review by D3DeFi — APPROVED** [review]
Looks good, thank you! :)

**Comment by D3DeFi** [comment]
shipit

**Comment by D3DeFi** [comment]
bot_status

**Comment by D3DeFi** [comment]
completely ignoring my shipit vote :( (mb cos PR was created before I've become maintainer)

It is possible that you will need to close & reopen the PR if this will not help (which I think will not :) )
!needs_revision

**Comment by D3DeFi** [comment]
ok, so my last idea is to close & reopen, can you do that please @Akint ?

**Comment by Akint** [comment]
> ok, so my last idea is to close & reopen, can you do that please @Akint ?

No problem, will do.

**Comment by D3DeFi** [comment]
!needs_revision
shipit

**Comment by D3DeFi** [comment]
!needs_revision

**Comment by D3DeFi** [comment]
shipit

**Comment by D3DeFi** [comment]
ready_for_review

**Comment by D3DeFi** [comment]
bot_status

**Comment by D3DeFi** [comment]
oh boy it ignores everything I try. @resmo care to step in please? I haven't seen @eikef in a while now and you've helped me in the past a few times (maybe I can use your help once again?). If I get your attention can you shipit/merge PR 45229 as well? (I don't want to link it as it is not related to this one)

btw is there some way for me to really get these PRs merged? Most of the time I am the only person "voting"

**Comment by resmo** [comment]
@D3DeFi sure, no problem, looking at it asap.

**Comment by resmo** [comment]
rebuild_merge

**Comment by resmo** [comment]
Thanks all for your works!

**Pull request merged**